beacon_node/http_api/src/ui.rs

```diff
@@ -0,0 +1,124 @@
+use beacon_chain::{metrics, BeaconChain, BeaconChainError, BeaconChainTypes};
+use eth2::types::ValidatorStatus;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::sync::Arc;
+use warp_utils::reject::beacon_chain_error;
+
+#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize)]
+pub struct ValidatorCountResponse {
+    pub active_ongoing: u64,
+    pub active_exiting: u64,
+    pub active_slashed: u64,
+    pub pending_initialized: u64,
+    pub pending_queued: u64,
+    pub withdrawal_possible: u64,
+    pub withdrawal_done: u64,
+    pub exited_unslashed: u64,
+    pub exited_slashed: u64,
+}
+
+pub fn get_validator_count<T: BeaconChainTypes>(
+    chain: Arc<BeaconChain<T>>,
+) -> Result<ValidatorCountResponse, warp::Rejection> {
+    let spec = &chain.spec;
+    let mut active_ongoing = 0;
+    let mut active_exiting = 0;
+    let mut active_slashed = 0;
+    let mut pending_initialized = 0;
+    let mut pending_queued = 0;
+    let mut withdrawal_possible = 0;
+    let mut withdrawal_done = 0;
+    let mut exited_unslashed = 0;
+    let mut exited_slashed = 0;
+
+    chain
+        .with_head(|head| {
+            let state = &head.beacon_state;
+            let epoch = state.current_epoch();
+            for validator in state.validators() {
+                let status =
+                    ValidatorStatus::from_validator(validator, epoch, spec.far_future_epoch);
+
+                match status {
+                    ValidatorStatus::ActiveOngoing => active_ongoing += 1,
+                    ValidatorStatus::ActiveExiting => active_exiting += 1,
+                    ValidatorStatus::ActiveSlashed => active_slashed += 1,
+                    ValidatorStatus::PendingInitialized => pending_initialized += 1,
+                    ValidatorStatus::PendingQueued => pending_queued += 1,
+                    ValidatorStatus::WithdrawalPossible => withdrawal_possible += 1,
+                    ValidatorStatus::WithdrawalDone => withdrawal_done += 1,
+                    ValidatorStatus::ExitedUnslashed => exited_unslashed += 1,
+                    ValidatorStatus::ExitedSlashed => exited_slashed += 1,
+                    // Since we are not invoking `superset`, all other variants will be 0.
+                    _ => (),
+                }
+            }
+            Ok::<(), BeaconChainError>(())
+        })
+        .map_err(beacon_chain_error)?;
+
+    Ok(ValidatorCountResponse {
+        active_ongoing,
+        active_exiting,
+        active_slashed,
+        pending_initialized,
+        pending_queued,
+        withdrawal_possible,
+        withdrawal_done,
+        exited_unslashed,
+        exited_slashed,
+    })
+}
+
+#[derive(PartialEq, Serialize, Deserialize)]
+pub struct ValidatorMetrics {
+    attestation_hits: u64,
+    attestation_misses: u64,
+    head_hit_percentage: f64,
```

Comment:
I think we may as well expose a few more quantities here, and then we can see what works best in the UI or have different values displayed in different places.

Good candidates for inclusion IMO:

- `attestation_head_hits`: based on `VALIDATOR_MONITOR_PREV_EPOCH_ON_CHAIN_HEAD_ATTESTER_HIT`
- `attestation_head_miss`: based on `VALIDATOR_MONITOR_PREV_EPOCH_ON_CHAIN_HEAD_ATTESTER_MISS`
- `attestation_target_hits`
- `attestation_target_miss`
- `hit_percentage`: equivalent to current impl of `head_hit_percentage`, i.e. `100 * attestation_hits / (attestation_hits + attestation_misses)`
- `head_hit_percentage`: re-defined in terms of `head` hit rate
- `target_hit_percentage`

I think that's probably good coverage for now, and we can come back to add more if we need later, e.g. aggregation delays, etc.

Reply:
Yes it does work without the `--metrics` flag.

Good point on the `head_hit_percentage`, I realize now it's not using the correct metrics, I'll fix that up and add all the others you suggested. :+1: